### ISSUE TITLE
Request: output doc for ArgumentCollection

### ISSUE THREAD

**Issue by magicDGS** [broadinstitute/barclay#7]
Argument collection could set documentation for all the options under the collection. I expect this to be printed out in the command line.

**Comment by cmnbroad** [comment]
The usage output is currently grouped by category (required, optional, etc.), and then listed alphabetically within each group. Although I can see some cases where using @ArgumentCollection the way you describe would be useful, in most cases I think its used to reflect code organization (and we're increasingly using it this way to allow sets of arguments to be used in multiple tool types, i.e., Spark and Walker). I think breaking up the output grouping by argument collection would introduce a lot of groups that may not be meaningful to the user.

 Anyone else see it differently ?

**Comment by magicDGS** [comment]
I agree with you to some extend. But for instance, if I have a group of arguments as this one:

```java
...
@ArgumentCollection(doc="Group of arguments for a concrete algorithm")
public void MyArgumentCollection args = new MyArgumentCollection();
...
```

I think that it makes sense to output in the command line the documentation for the group and group them together. Maybe if there is a check for the content of "doc", it could work in your use case (that I also like from the development point of group) and mine if while printing the help. For instance, if the doc is not empty, the arguments should be kept together and the doc should be printed out (similar to the argument plugins), and if not they are included in the alphabetical ordering.

**Comment by magicDGS** [comment]
@cmnbroad, what about an optional boolean in `@ArgumentCollection` to keep them together and output the doc? I will offer myself to do this...

**Comment by cmnbroad** [comment]
I'm not sure what the right solution is. It seems wrong to me to have the doc and grouping attributes be on the ArgumentCollection annotation, since then the values have to be specified at each call-site (which seems redundant and error prone). 

BTW, I just noticed that the command line usage output currently doesn't distinguish between tool-args and common args. I made a simple PR to fix this since it seems like it gets us part way to what we're talking about here (though currently not all of the common args in GATK are annotated correctly, but thats a separate issue). So that change alone would mean that on the command line, you'd see args grouped like this:

Required Args
Optional tool-specific args
Optional common args

So in the case of, say, ApplyBQSR, you'd see all of the BQSR-related args grouped together and separate from the "common" args like readfilters, intervals, etc. (at least once we fix the annotations). The online doc gen already separates them like this.

I wonder if that sufficient. And if its not, we might want to consider something targeted to online output so we don't overcrowd the usage output.

**Comment by cmnbroad** [comment]
I see the value in having collection-level doc, but I think it introduces complexity as well as  conflicts with the current "usage output" and "online doc" models (i.e., if an argument collection with doc has a required arg, do we list that arg under the arg collection heading; under required; or under both ?). Same for online doc (if we add it to the annotation we'd have to integrate it with Barclay docgen). @lbergelson Any thoughts on this ?

**Comment by magicDGS** [comment]
I see also the complications of `ArgumentCollection` with doc and required arguments. I think that in this situation it should blow up (the same as with `@Hidden` and required arguments). How the doc for plugins with required arguments is handled?

**Comment by cmnbroad** [comment]
Required arguments in plugins are not really required; they're only "conditionally" required if the containing plugin is specified on the command line. For example, ReadLengthReadFilter has a required arg, but its only required if the user actually specifies ReadLengthReadFilter on the command line.

In usage output, all of the args for a given plugin show up at the end, grouped together under the parent plugin. For online doc, each plugin has it's own doc page (this is only by convention), so they tool doc just includes a link to that page.

**Comment by magicDGS** [comment]
Ok, thanks for the explanation :)
